clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -979,12 +1026,124 @@ private CompletableFuture<Void> invokeOnPartitionsLostCallback(Set<TopicPartitio
         // behaviour.
         Optional<ConsumerRebalanceListener> listener = subscriptions.rebalanceListener();
         if (!partitionsLost.isEmpty() && listener.isPresent()) {
-            throw new UnsupportedOperationException("User-defined callbacks not supported yet");
+            return enqueueConsumerRebalanceListenerCallback(onPartitionsLost, partitionsLost);
         } else {
             return CompletableFuture.completedFuture(null);
         }
     }
 
+    /**
+     * Enqueue a {@link ConsumerRebalanceListenerCallbackNeededEvent} to trigger the execution of the
+     * appropriate {@link ConsumerRebalanceListener} {@link ConsumerRebalanceListenerMethodName method} on the
+     * application thread.
+     *
+     * <p/>
+     *
+     * This method is essentially "giving" the baton from the background thread to the application thread for
+     * processing of the reconciliation logic. It will "receive" the "baton" back via the
+     * {@link #consumerRebalanceListenerCallbackCompleted(ConsumerRebalanceListenerMethodName, Optional)} method.
+     *
+     * <p/>
+     *
+     * Because the reconciliation process (run in the background thread) will be blocked by the application thread
+     * until it completes this, we need to leave a {@link ConsumerRebalanceListenerCallbackBreadcrumb breadcrumb}
+     * by which to remember where we left off.
+     *
+     * @param methodName Callback method that needs to be executed on the application thread
+     * @param partitions Partitions to supply to the callback method
+     * @return Future that will be chained within the rest of the reconciliation logic
+     */
+    private CompletableFuture<Void> enqueueConsumerRebalanceListenerCallback(ConsumerRebalanceListenerMethodName methodName,
+                                                                             Set<TopicPartition> partitions) {
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        ConsumerRebalanceListenerCallbackBreadcrumb newBreadcrumb = new ConsumerRebalanceListenerCallbackBreadcrumb(
+            methodName,
+            future
+        );
+
+        if (breadcrumbRef.compareAndSet(null, newBreadcrumb)) {
```

Comment:
I have doubts about this. My understanding is that you are trying to prevent the state machine from scheduling multiple callbacks. Is my understanding right?

For the context, I think that the current implementation of the state machine can actually schedule multiple callbacks. It does not do it in the reconciliation process but it could happen when the member is fenced or when the user leaves/unsubscribe. So, I think that this won't work. @lianetm It would be great if you could look into this as well.

Reply:
Agree. During the reconciliation the state machine ensures that only one callback is triggered at a time. But member could leave (triggers `onPartitionsRevoked`), or get fenced (triggering `onPartitionsLost`) at any time (ex. during a reconciliation). Thus we could have multiple callbacks being executed at the same time.